### ISSUE TITLE
Asking for help on the extension of the WYSIWYG rendering formula of tui.editor

### ISSUE THREAD

**Issue by jack9603301** [nhn/tui.editor#486]
Because I need to solve the problem that the WYSIWYG editor can not convert normally when it parses mathematical formulas similar to the following:
```markdown
\(\sqrt{3}\)
\[\sqrt{3}\]
```
Our plug-in supports the parsing of plug-ins and non-code blocks using code block mode. At present, the Review and rendered documents have been transformed and parsed directly through plug-ins, but it still can not solve the normal rendering of WYSIWYG editor. Therefore, we intend to re-encapsulate the converter and replace it with an extended way.

But after trying, we found that we could not get the built-in converter of tui.editor directly. We asked for help and told us what we should do. Or is there another way to solve this problem perfectly? Do you need to customize the tui.editor code?

**Comment by sohee-lee7** [comment]
@jack9603301 Sorry for late reply. I didn't understand your problem. Is it the problem for you that `(sqrt{3})` convert to `\(\sqrt{3}\)` with slash? 

Built-in converter is not opened but you can use events that occur before and after convert.

- `convertorAfterMarkdownToHtmlConverted`
- `convertorBeforeHtmlToMarkdownConverted`
- `convertorAfterHtmlToMarkdownConverted`

I hope those event help you. If you still have problems, please leave a comment again.

**Comment by jack9603301** [comment]
I need to parse the latex standard grammar of mathematical formulas and render them by calling katex, etc.

**Comment by qiujiangkun** [comment]
What we actually need is the API by which we can add our own extension rendering math formula in WYSIWYG mode.

**Comment by xland** [comment]
KaTeX ‘s API   https://katex.org/

**Comment by JonasOelke** [comment]
Hey Guys, 
I think I'm trying the same you do. @jack9603301 Did you already solve the problem? 
By now I've got a separate div where my Latex will be displayed rendered:

```
editor.on('change', function() {
            if (document.getElementById("LatexPreview").style.display != "none") {
                document.getElementById("LatexPreview").innerHTML = editor.getHtml(); 
                renderMathInElement(document.getElementById("LatexPreview"))
            }
        });
```

This is what's working for now, but it's not really nice ... 

Atm I'm trying the following: 

```
editor.on('change', function() {
            renderMathInElement(document.getElementsByClassName("tui-editor-contents")[0])
        });
```

renderMathInElement is a Katex function, it get's a DOM-element looks for all LaTeX in it and renders it. This does nothing in wysiwyg mode and in markdown mode you see the rendered LaTeX for about a second in the preview section but then it disappears and you still only see the LaTeX Code.

I probably can't just render the LaTeX in the "tui-editor-contents" section. But I don't know where I can kinda add some commands into the _render engine_ ?! Any Ideas? Is there a Plugin API for rendering extensions? Tbh there isn't a really helpful documentation on this, or at least I can't find anything helpful. 

Jonas

**Comment by qiujiangkun** [comment]
Hi Jonas Oelke,
I was working on a same project with Jack. Until now we found markdownit a proper place to set up katex as a markdownit plugin.
Though it's kinda ugly.
There're still tons of work to do, such as different encoding, etc. But it worked.
Check our project out: Wp reliable md.
Best wishes

**Comment by JonasOelke** [comment]
Hi @qiujiangkun, 

there is already a Latex Plugin for Markdown: https://github.com/yzhang-gh/markdown-it-katex

To be honest I'm quite new to web development and I have no clue where I could start to implement this markdownit plugin to the tui.editor - it seems like there is tons of code lines to understand before being able to tell where and how to implement it. Even the usage explained in the plugin's repository looks quite easy. 

I will have a look at your project - even though it's a Wordpress Plugin, I might be able to find a way to implement it in out node.js project. 

Can't actually imagine that it's so tricky to implement such an easy to use render tool. Maybe one tui.editor's developer guys can tell what might be the easiest way. 

But thank you anyway! 
Jonas

**Comment by qiujiangkun** [comment]
Hello, JonasOelke,
We prepared our own plugin to support Katex and it works. We did not use Katex autorender but manually render every formula.
I guess these js files will help you. Just take a look at them.

https://github.com/jackworkshop/WP-ReliableMD/tree/master/js
https://github.com/jackworkshop/WP-ReliableMD/blob/master/js/TuiMathSupport.js
https://github.com/jackworkshop/WP-ReliableMD/blob/master/js/TuiEditorMathSupport.js
https://github.com/jackworkshop/WP-ReliableMD/blob/master/js/TuiViewerMathSupport.js

**Comment by jack9603301** [comment]
I started a project with @qiujiangkun, which plans to allow users to choose latex math rendering engine on their own, so we need Tui to support both KAtex and mathjax. Our plug-in can work properly with the final result, but we can't let the edit-time preview window work. Maybe later, we will add other functions to the plug-in, so we hope tui's editor engine has enough extensibility, especially KAtex and mathjax. Some of the responses on @qiujiangkun may help you.

**Comment by jack9603301** [comment]
@JonasOelke , I have not solved this problem, because I have some things at present, so there is only one future development plan and idea for this problem. We have adopted two strategies:

1. Because I and another development partner have some things that have delayed Wordpress Markdown's plug-in development plan, we first submitted issus in the hope of seeking official help. We sincerely hope that Tui can do better.

2. If the official can't solve the problem well, when we have time, we will modify the Tui editor kernel directly.
